File: src/datasource.js
/* This file should export the following functions
    async function initDatasource() //May throw error.
    async function search() //May throw error.
 */

const { escapeRegExp, sortBookPrice, sortBookPriceDesc } = require('./utils')
//All data shall be loaded in this variable.
var data = []


// This function is on the start of the application to create the connection with
// the database (in this case, with the books.json file.).
module.exports.initDatasource = async function() {
    console.log('Starting initiating datasouce')

    const fs = require('fs')
    const util = require('util')

    const readfile = util.promisify(fs.readFile)

    var booksString = await readfile('./books.json')

    data = JSON.parse(booksString)

    console.log('Successfully initiated datasource')
    return
}


module.exports.search = async function(params) {

    // The script here defined is not the best for large datasets, therefore,
    // it is recomended using a database with fast search, using for example 
    // Lucene ElasticSearch

    var res = data


    //Searches the name
    if (params.Name) {
        var nameRegex = new RegExp(escapeRegExp(params.Name), 'i')
        res = res.filter(book => nameRegex.test(book.name))
        delete params.Name
    }

    var sort = ''
    //Gets sort param
    if (params.Sort) {
        sort = params.Sort
        delete params.Sort
    }

    //Searches the rest of the inputs
    for (var key in params) {
        var paramRegex = new RegExp(escapeRegExp(params[key]), 'i')

        res = res.filter(book => {

            if (book.specifications[key]) { // Key exists 
                var testArray = book.specifications[key];

                if (!Array.isArray(book.specifications[key])) { // Transform into an array of strings...
                    testArray = [book.specifications[key].toString()]
                }
                // Check if at least one element contain the string specified
                return testArray.some(
                    specification => paramRegex.test(specification))
            }
            else { // Book does not contain the specification specified...
                return false
            }
        })
    }

    if (sort === 'asc') {
        res = res.sort(sortBookPrice)
    }
    if (sort === 'desc') {
        res = res.sort(sortBookPriceDesc)
    }

    return res
}

File: src/datasource.rb
require 'json'

# Datasource
module Datasource
  # Read the Json file and save the data in a Hash
  def self.init
    @data_hash = {}
    @params = {}
    print("---Reading Json file \n")
    file = File.read('./books.json')
    @data_hash = JSON.parse(file)

    @data_hash
  end

  # Search for book name passed in URL

  def self.search_params_name(answer, params)
    return answer if params['Name'].nil?

    answer = answer.select { |book| book['name'].include? params['Name'] }
    params.delete('Name')
    answer
  end

  # Save the order parameter

  def self.search_params_sort(params)
    return if params['Sort'].nil?

    sort = params['Sort']
    params.delete('Sort')
    sort
  end

  # Search for specifications, use the params passed in URL. Check if exist in
  # json file then return a response

  def self.search_params_specifications(answer, params)
    params.each do |key, value|
      current_key = key.capitalize
      next unless answer.first['specifications'].key?(current_key.to_s)

      #convert hash to string and search if have substring value
      answer = answer.select { |book| book['specifications'][current_key].to_s.include? value }
    end
    answer
  end

  # Main function of search
  def self.search(params)
    answer = @data_hash
    @params = params

    answer = search_params_name(answer, @params)
    sort = search_params_sort(@params)
    answer = search_params_specifications(answer, @params)

    return answer.sort_by { |book| book['price'] } if sort == 'asc'
    return answer.sort_by { |book| -book['price'] } if sort == 'desc'

    answer
  end
end
